File: cpp/s0003.cc
#include "essential.h"
#include <unordered_set>

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    int len = s.size();
    unordered_set<char> st;
    int start(0), end(0), maxSize(0);
    while (end < len) {
      if (st.count(s[end]) == 0) {
        st.insert(s[end]);
        maxSize = max(maxSize, (int)st.size());
        end++;
      } else {
        st.erase(s[end + 1]);
        start++;
      }
    }
    return maxSize;
  }
};

int main() { return 0; }

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    unordered_set<char> set;
    int left = 0, max_len = 0;
    for (int r = 0; r < s.size(); r++) {
      while (set.count(s[r])) {
        set.erase(s[left]);
        left += 1;
      }
      set.insert(s[r]);
      max_len = max(max_len, r - left + 1);
    }
    return max_len;
  }
};

File: rust/src/s0003.rs
pub struct Solution {}

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        use std::collections::HashSet;
        let mut set: HashSet<char> = HashSet::new();
        let (mut i, mut j) = (0, 0);
        let mut result = 0;
        let s = s.chars().into_iter().collect::<Vec<char>>();
        while j < s.len() {
            if !set.contains(&s[j]) {
                set.insert(s[j]);
                result = std::cmp::max(result, j - i + 1);
                j += 1;
            } else {
                set.remove(&s[i]);
                i += 1;
            }
        }
        result as i32
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        assert_eq!(
            Solution::length_of_longest_substring("abcabcbb".to_string()),
            3
        );
    }
}
